src/audio/asrc/asrc_farrow.c

```diff
@@ -0,0 +1,1398 @@
+	src_obj->prim_num_frames_targ = 0;
+	src_obj->sec_num_frames = 0;
+	src_obj->sec_num_frames_targ = 0;
+	src_obj->calc_ir = NULL;
+	src_obj->is_initialised = false;
+	src_obj->is_updated = false;
+
+	/* check conversion ratios */
+	if (src_obj->fs_ratio == 0 || src_obj->fs_ratio_inv == 0)
+		return ASRC_EC_INVALID_CONVERSION_RATIO;
+
+	/*
+	 * The pointer to the internal ring buffer pointers is
+	 * pointing to the memory subsequently to the memory where the
+	 * src_obj lies.  Only one of the buffers is initialised,
+	 * depending on the specified bit depth.
+	 */
+	if (src_obj->bit_depth == 32) {
+		src_obj->ring_buffers32 = (int32_t **)(src_obj + 1);
+
+		/* error check */
+		if (!src_obj->ring_buffers32)
+			return ASRC_EC_INVALID_BUFFER_POINTER;
+
+		src_obj->ring_buffers16 = NULL;
+	} else if (src_obj->bit_depth == 16) {
+		src_obj->ring_buffers16 = (int16_t **)(src_obj + 1);
+
+		/* error check */
+		if (!src_obj->ring_buffers16)
+			return ASRC_EC_INVALID_BUFFER_POINTER;
+
+		src_obj->ring_buffers32 = NULL;
+	}
+
+	/*
+	 * Load the filter coefficients and parameters.  This function
+	 * also sets the pointer to the corresponding
+	 * calc_impulse_response_nX function.
+	 */
+	error_code = initialise_filter(src_obj);
+
+	/* check for errors */
+	if (error_code != ASRC_EC_OK)
+		return error_code;
+
+	/* set the channel pointers and fill buffers with zeros */
+	error_code = initialise_buffer(src_obj);
+
+	/* check for errors */
+	if (error_code != ASRC_EC_OK)
+		return error_code;
+
+	/*
+	 * Set the pointer for the impulse response.
+	 * &m_impulse_response[0] = ring_buffers_x
+	 * + num_channels * (1 + buffer_length);
+	 * Here ring_buffer_x already points to the memory just behind
+	 * the src_obj.  The offset results from the number of pointers
+	 * pointing to each channel and each channels buffer data.
+	 */
+	if (src_obj->bit_depth == 32) {
+		src_obj->impulse_response =
+			(int32_t *)(src_obj->ring_buffers32 +
+				    src_obj->num_channels *
+				    (1 + src_obj->buffer_length));
+	} else if (src_obj->bit_depth == 16) {
+		src_obj->impulse_response =
+			(int32_t *)(src_obj->ring_buffers16 +
+				    src_obj->num_channels *
+				    (1 + src_obj->buffer_length / 2));
+	}
+
+	/* check for invalid pointer */
+	if (!src_obj->impulse_response)
+		return ASRC_EC_INVALID_POINTER;
+
+	/* return ok, if everything worked out */
+	src_obj->is_initialised = true;
+	return ASRC_EC_OK;
+}
+
+enum asrc_error_code set_fs_ratio(struct asrc_farrow *src_obj, int32_t fs_prim,
+				  int32_t fs_sec)
+{
+	/* Check for parameter errors */
+	if (!src_obj)
+		return ASRC_EC_INVALID_POINTER;
+
+	if (!src_obj->is_initialised)
+		return ASRC_EC_INIT_FAILED;
+
+	if (fs_prim < 8000 || fs_prim > 192000)
+		return ASRC_EC_INVALID_SAMPLE_RATE;
+
+	if (fs_sec < 8000 || fs_sec > 192000)
+		return ASRC_EC_INVALID_SAMPLE_RATE;
+
+	/* set sampling rates */
+	src_obj->fs_prim = fs_prim;
+	src_obj->fs_sec = fs_sec;
+
+	/* calculate conversion ratios as Q5.27 */
```

Comment:
This part of the code is a bit awkward. A normal signed  Q5.27 is up to 15.9999... so it could not handle 192/8 = 24. However the storage for ratios is uint32_t so the up to 31.9999... range is sufficient. I dislike uint32_t arithmetic because the DSP can't use the the efficient fractional maths instructions for them. I would have myself used Q6.26 instead (26 decimal bits is still a very good accuracy) but I didn't want to make changes to algo core in this point.

I'll add a comment  about the ranges to avoid confusion.

Reply:
done!